### PR TITLE
feat: gateway controller

### PR DESCRIPTION
**Pull request by shaneutt**
**What this PR does / why we need it**:

This adds a `Gateway` controller as the first stage of [Gateway APIs](https://github.com/kubernetes-sigs/gateway-api) support in the KIC.

**Which issue this PR fixes**

Resolves https://github.com/Kong/kubernetes-ingress-controller/issues/1945

**Special notes for your reviewer**:

This iteration doesn't provide any end-user functionality, and is simply the first iteration in [Gateway Milestone 1](https://github.com/Kong/kubernetes-ingress-controller/milestone/21).

As per the [KEP](https://github.com/Kong/kubernetes-ingress-controller/blob/main/keps/0002-kong-gateway-api.md) this iteration is limited to "pre-provisioned" mode, meaning an existing Kong Gateway must exist in order to "provision" `Gateway` resources.

**PR Readiness Checklist**:

- [x] the `CHANGELOG.md` release notes have been updated

### CODE REVIEW

File: internal/controllers/gateway/gateway_controller.go [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+			}
+		}
+	}
+
+	return addresses, listeners, nil
+}
+
+// determineL6and7ListenersFromDataPlane takes a list of L4 Gateway listeners and references
+// them against the data-plane to determine any L6 (TLS) or L7 (HTTP, e.t.c.) support configured for them.
+func (r *GatewayReconciler) determineL6and7ListenersFromDataPlane(ctx context.Context, svc *corev1.Service, listeners []gatewayv1alpha2.Listener) ([]gatewayv1alpha2.Listener, error) {
```
Comment (mflendrich) [inline internal/controllers/gateway/gateway_controller.go:495]: I don't see why TLS would be L6 and not L5.

In general the OSI layer metaphore doesn't stick with me, but as long as it's explained clearly that's fine.

This is not a request for change, just inspiration: my take on the nomenclature here would be something like replacing:
- "L4" and "L6" with "HostPort" and "Stream"
- "L7" with "RequestProxy" or something similar.

File: internal/controllers/gateway/gateway_controller.go [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+	return upgradedListeners, nil
+}
+
+// -----------------------------------------------------------------------------
+// Gateway Controller - Private Object Update Methods
+// -----------------------------------------------------------------------------
+
+// updateAddressesAndListenersSpec updates a unmanaged gateway's spec with new addresses and listeners.
+// If the addresses and listeners provided are the same as what exists, the update is skipped.
+func (r *GatewayReconciler) updateAddressesAndListenersSpec(ctx context.Context, gateway *gatewayv1alpha2.Gateway, addresses []gatewayv1alpha2.GatewayAddress, listeners []gatewayv1alpha2.Listener) (bool, error) {
```
Comment (mflendrich) [inline internal/controllers/gateway/gateway_controller.go:547]: this logic is not unit-testable and I think it should be unit-tested

File: internal/manager/controllerdef.go [hunk 3]
```diff
@@ -217,6 +220,27 @@ func setupControllers(mgr manager.Manager, proxy proxy.Proxy, c *Config, feature
...
+					Version:  gatewayv1alpha2.SchemeGroupVersion.Version,
+					Resource: "gateways",
+				}}.CRDExists,
+			Controller: &gateway.GatewayReconciler{
+				Client:          mgr.GetClient(),
+				Log:             ctrl.Log.WithName("controllers").WithName("Gateway"),
+				Scheme:          mgr.GetScheme(),
+				Proxy:           proxy,
+				PublishService:  c.PublishService,
+				KongConfig:      kongConfig,
```
Comment (mflendrich) [inline internal/manager/controllerdef.go:240]: IMO including the whole of Kongconfig is too broad and hurts cohesion. I suggest injecting just a listener getter function, or the list of listeners read just once upon startup.

**Review by mflendrich — COMMENTED** [review]
Looks good. I like the approach.

Comments centered mainly around future maintainability and testability.

File: internal/controllers/gateway/gateway_controller.go [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+	// due to a Gateway that is not properly configured for unmanaged mode.
+	ManagedGatewaysUnsupported = fmt.Errorf("invalid gateway spec: managed gateways are not currently supported") //nolint:revive
+)
+
+// -----------------------------------------------------------------------------
+// Gateway Controller - GatewayReconciler
+// -----------------------------------------------------------------------------
+
+// GatewayReconciler reconciles a Gateway object
+type GatewayReconciler struct { //nolint:revive
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go:56]: To allow a public symbol prefixed with the same name as the package.

File: internal/controllers/gateway/gateway_controller.go [hunk 5]
```diff
@@ -0,0 +1,581 @@
...
+		return ctrl.Result{}, r.Status().Update(ctx, pruneGatewayStatusConds(gateway))
+	}
+
+	// allow for Gateway resources to be configured with "true" in place of the publish service
+	// reference as a placeholder to automatically populate the annotation with the namespace/name
+	// that was provided to the controller manager via --publish-service.
+	debug(log, gateway, "initializing admin service annotation if unset")
+	if existingGatewayEnabled == "true" { // true is a placeholder which triggers auto-initialization of the ref
+		debug(log, gateway, fmt.Sprintf("a placeholder value was provided for %s, adding the default service ref %s", UnmanagedAnnotation, r.PublishService))
+		gateway.Annotations[UnmanagedAnnotation] = r.PublishService
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go:307]: Can you please point me to an example setter method, I appear to be missing them.

File: internal/controllers/gateway/gateway_controller.go [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+	// support connected Gateway resources.
+	ControllerName gatewayv1alpha2.GatewayController = "konghq.com/kic-gateway-controller"
+
+	// UnmanagedAnnotation is an annotation used on a Gateway resource to
+	// indicate that the Gateway should be reconciled according to unmanaged
+	// mode.
+	//
+	// NOTE: it's currently required that this annotation be present on all Gateway
+	// resources: "unmanaged" mode is the only supported mode at this time.
+	UnmanagedAnnotation = "konghq.com/gateway-unmanaged"
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go]: dd041b93200c82f6361ae8e487271d5865cd49e6

File: internal/controllers/gateway/gateway_controller.go [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+
+// reconcileUnmanagedGateway reconciles a Gateway that is configured for unmanaged mode,
+// this mode will extract the Addresses and Listeners for the Gateway from the Kubernetes Service
+// used for the Kong Gateway in the pre-existing deployment.
+func (r *GatewayReconciler) reconcileUnmanagedGateway(ctx context.Context, log logr.Logger, gateway *gatewayv1alpha2.Gateway) (ctrl.Result, error) {
+	// currently this controller supports only unmanaged gateway mode, we need to verify
+	// any Gateway object that comes to us is configured appropriately, and if not reject it
+	// with a clear status condition and message.
+	debug(log, gateway, "validating management mode for gateway") // this will also be done by the validating webhook, this is a fallback
+	existingGatewayEnabled, ok := gateway.GetAnnotations()[UnmanagedAnnotation]
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go]: fab8687666439588a36b43b2fa126f5a33cfe205

File: internal/controllers/gateway/gateway_controller.go [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+	if !ok || existingGatewayEnabled == "" {
+		log.Error(ManagedGatewaysUnsupported, fmt.Sprintf("%s=namespace/name annotation is currently required, will not retry", UnmanagedAnnotation), "namespace", gateway.Namespace, "name", gateway.Name)
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go]: fb39090fd1987de00de0b659fbf8808fa331ee8d

File: internal/controllers/gateway/gateway_controller.go [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+		gateway.Status.Conditions = append(gateway.Status.Conditions, metav1.Condition{
+			Type:               string(gatewayv1alpha2.GatewayConditionScheduled),
+			Status:             metav1.ConditionFalse,
+			ObservedGeneration: gateway.Generation,
+			LastTransitionTime: metav1.Now(),
+			Reason:             string(gatewayv1alpha2.GatewayReasonNoResources),
+			Message:            "managed gateways are not currently suppported, this object can not be scheduled",
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go]: cefce4e60220a290bff9b29629a00063454df665

File: internal/controllers/gateway/gateway_controller.go [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+		})
+		return ctrl.Result{}, r.Status().Update(ctx, pruneGatewayStatusConds(gateway))
+	}
+
+	// allow for Gateway resources to be configured with "true" in place of the publish service
+	// reference as a placeholder to automatically populate the annotation with the namespace/name
+	// that was provided to the controller manager via --publish-service.
+	debug(log, gateway, "initializing admin service annotation if unset")
+	if existingGatewayEnabled == "true" { // true is a placeholder which triggers auto-initialization of the ref
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go:299]: fab8687666439588a36b43b2fa126f5a33cfe205

File: internal/controllers/gateway/gateway_controller.go (hunk 1 above)
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go]: This is I think a source of debate to this day, so rather than worry about it I've changed the language to not reference the model 17392644e3d9ef49998badb47cc385f56da92370

File: internal/manager/controllerdef.go (hunk 3 above)
Comment (shaneutt) [reply internal/manager/controllerdef.go]: The `Gateway` resource is meant to be a living reflection of the Kong Gateway itself, and needs access to the data-plane in order to make that reflection work. I'm really not seeing how having controllers in our control plane which is entirely about this data-plane having access to that data-plane's API to be too broad or not cohesive. We do however usually wrap these kinds of things behind the `Proxy` interface which was purpose built to be such an abstraction, so I moved the logic under there in 4d0023108da415a2bc842794f5c8af63a63fbd55 let me know if that seems sufficient.

File: internal/controllers/gateway/gateway_controller.go (hunk 2 above)
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go:542]: The integration tests cover this logic, and the logic is bound up in the reconcilation loop pretty strongly. Can you help me to understand better what's missing and perhaps we can simply add some more integration tests to cover it?

File: internal/controllers/gateway/gateway_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mflendrich) [reply internal/controllers/gateway/gateway_controller.go]: There are none FWIW. It's the first time we're implementing a setter method, and I thought it makes sense for one to live there.

File: internal/controllers/gateway/gateway_controller.go (hunk 2 above)
Comment (mflendrich) [reply internal/controllers/gateway/gateway_controller.go:542]: We have a chunk of logic that performs reflection magic. This is opaque and understandign that it does _exactly_ what it's intended to is IMO desirable. It's easy to lose this level of detail (specifically: that we use an arcane mechanism to do a simple thing) on the integration testing level.

What I'm looking for is a table-based unit test that has `gateway` `addresses`, `listeners` as inputs, and `true/false` and the resulting `gateway` as expected outputs.

File: internal/controllers/gateway/gateway_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go]: Ok, then it seems to me that the thing to do for consistency would be just to set the annotation value as normal?

File: internal/manager/controllerdef.go (hunk 3 above)
Comment (mflendrich) [reply internal/manager/controllerdef.go]: Yes this looks reasonable :+1:

File: internal/controllers/gateway/gateway_controller.go (hunk 2 above)
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go:542]: I'm having a hard time seeing `DeepEquals` as reflection magic. I think one could argue the function could be moved out of reflection because it's kind of a weird place for it to be, but that's neither here nor there. It's a very common tool used to compare the equality of nested structures. It sounds like we're asking to unit test `DeepEquals` which seems out of scope for this work.

Comment (mflendrich) [reply internal/controllers/gateway/gateway_controller.go:542]: > It sounds like we're asking to unit test DeepEquals

We're asking to unit test a logical conjunction of two usages of reflection to perform comparison (which is surely unit tested, but could potentially be prone to corner cases) and a resulting decision.

The fact that we use an external library (reflect) instead of simple comparisons makes this even more reasonable to unit test IMO.

Comment (mflendrich) [reply internal/controllers/gateway/gateway_controller.go:542]: To invert the question a little bit:

- Is the purpose of unit testing to detect corner case behavior and logic bugs?
- If the answer to the above question is yes: what differentiates this chunk of logic from another (hypothetical) chunk of logic that would deserve an unit test?

File: internal/controllers/gateway/gateway_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mflendrich) [reply internal/controllers/gateway/gateway_controller.go]: What KIC does before this change is that it uses wrappers from `pkackage annotations` to access annotations instead of manipulating them directly.

By not having a setter function we'd have an inconsistent abstraction: reads happening indirectly, but writes happening directly.

Does this rationale make sense to you?

[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go]: Maybe but it just seems like I'm going to start a new trend by doing that here and I don't see the value.

File: internal/controllers/gateway/gateway_controller.go (hunk 2 above)
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go:542]: We've always wanted to unit test everything, but we've settled for integration tests in the past. It sounds like this PR has become the duck in a game of duck duck goose for starting to re-apply pressure on that, which I suppose makes some sense anyway it's the first major change since 2.0. That said, I think we're missing the forest for the trees here: we need to be institutionalizing controller unit tests as opposed to picking out specific methods and intentionally wrapping them around sub-functions just so they can be partially testable. I've created https://github.com/Kong/kubernetes-ingress-controller/issues/2006 as a follow up to this, but I'll also do what is requested here and break this out into a unit-tested function in the meantime.

Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go:542]: After re-acquainting myself with `envtest` I remembered how heavyweight it is. I have a partial implementation in a branch now, but there's a _lot_ more work to do to get that done that I don't feel should block this. As such #2006 is the follow up, but I added unit tests for this in e9e436f18596b8c5f6945f31ed530a614fd3a860 as requested.

File: internal/controllers/gateway/gateway_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rainest) [reply internal/controllers/gateway/gateway_controller.go]: The existing read abstraction is kinda pointless. AFAICT we only have those functions because annotation values are always strings, and the Kong route/service/etc. we stuff them into use their actual type. For example, the getter for a bool Kong entity field converts `"true"` to `true`, a list field converts `"foo,bar"` to `[]string{"foo", "bar"}`, etc.

We never do the reverse and stuff Kong values back into an annotation, and don't usually set annotations at all outside of tests (this is, in fact, the only location we do it per `grep -r "\.Annotations\[" | grep -vi test`).

The getters are of fairly limited use themselves. We only use each once within the parser (`grep -o -r -P "annotations.Has.*\(.*\)" ./ | grep -o -P "annotations.Has.*\(.*\)" | sort | uniq -`) so they could be reasonably inlined. I'd say do that instead if you want consistency (every setter other than UnmanagedAnnotation's would be dead code as we have no use for them), but IMO not worth it either way.

**Review by rainest — APPROVED** [review]

**Review by rainest — APPROVED** [review]

File: internal/controllers/gateway/gateway_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/controllers/gateway/gateway_controller.go]: This comment is holding up the PR merge, since the original review was non-blocking I'm assuming it's OK to move forward for now. If you would like us to follow-up on this let me know and I'll be happy to.
